src/sst/core/interprocess/tunneldef.h

```diff
@@ -0,0 +1,262 @@
+     * @param command message to write to buffer
+     */
+    void writeMessage(size_t  buffer, const MsgType &command) {
+        circBuffs[buffer]->write(command);
+    }
+
+    /** Read data from buffer, blocks until message received
+     * @param buffer which buffer to read from
+     * return the message
+     */
+    MsgType readMessage(size_t buffer) {
+        return circBuffs[buffer]->read();
+    }
+
+    /** Read data from buffer, non-blocking
+     * @param buffer which buffer to read from
+     * @param result pointer to return read message at
+     * return whether a message was read
+     */
+    bool readMessageNB(size_t buffer, MsgType *result) {
+        return circBuffs[buffer]->readNB(result);
+    }
+
+    /** Empty the messages in a buffer
+     * @param buffer which buffer to empty
+     */
+    void clearBuffer(size_t buffer) {
+       circBuffs[buffer]->clearBuffer();
+    }
+
+    /** return whether this is a master-side tunnel or a child*/
+    bool isMaster() { 
+        return master; 
+    }
+
+private:
+    /** Allocate space for a data structure in the shared region
+     * @tparam T data structure type to allocate space for
+     * @param extraSpace how many extra bytes to reserve with this allocation
+     * return offset from shmPtr where structure was allocated and pointer to structure 
+     */
+    template <typename T>
+    std::pair<size_t, T*> reserveSpace(size_t extraSpace = 0)
+    {
+        size_t space = sizeof(T) + extraSpace;
+        if ( (size_t)((nextAllocPtr + space) - (uint8_t*)shmPtr) > shmSize )
+            return std::make_pair<size_t, T*>(0, NULL);
+        T* ptr = (T*)nextAllocPtr;
+        nextAllocPtr += space;
+        new (ptr) T();  // Call constructor if need be
+        return std::make_pair((uint8_t*)ptr - (uint8_t*)shmPtr, ptr);
+    }
+    
+    /** Calculate the size of the tunnel */
+    static size_t calculateShmemSize(size_t numBuffers, size_t bufferSize) {
+        long pagesize = sysconf(_SC_PAGESIZE);
+        /* Count how many pages are needed, at minimum */
+        size_t isd = 1 + ((sizeof(InternalSharedData) + (1+numBuffers)*sizeof(size_t)) / pagesize);
+        size_t buffer = 1 + ((sizeof(CircularBuffer<MsgType>) + bufferSize*sizeof(MsgType)) / pagesize);
+        size_t shdata = 1 + ((sizeof(ShareDataType) + sizeof(InternalSharedData)) / pagesize);
+
+        /* Alloc 2 extra pages just in case */
+        return (2 + isd + shdata + numBuffers*buffer) * pagesize;
+    }
+
+
+protected:
+    /** Pointer to the Shared Data Region */
+    ShareDataType *sharedData;
+
+    /** Return the number of buffers */
+    size_t getNumBuffers() { return numBuffs; }
+
+private:
+    bool master;
+    void *shmPtr;
+
+    uint8_t *nextAllocPtr;
+    size_t shmSize;
+
+    // Local data
+    size_t numBuffs;
+    size_t buffSize;
+    uint32_t children;
+
+    // Shared objects
+    InternalSharedData *isd;
+    std::vector<CircBuff_t* > circBuffs;
```

Comment:
I'm not sure we have ABI compatibility between C++11 and C++98. Does this get built with C++11 in SST and then not C++11 in PIN land?

Reply:
Yes. I also just pushed the ariel-side branch in sst-elements "ariel_pin3" which has all of the Pin stuff.

Reply:
We might never ever have a problem.... but.... if we could make this STL-free it would guarantee we don't have an impossible to find bug later. I'm basically 100% sure that pair would have no ABI issues. But, since the first type is also size_t, we could replace it with a `queue_entry<T>`.  I again doubt that `vector` would ever be an issue... but this seems far more likely to have ABI issues between 98 and 11 if we can just use a raw pointer array.

Reply:
That's a good point, I'll rework it.

Reply:
Looked at this some more since (it's copied from the original IPCTunnel) and I actually think it is OK. The vector itself lives in this class, not in the shared region. The vector is filled with pointers constructed using offsets stored in a plain array the shared region. The region contains the ISD struct, the shared data struct, and the circular buffers, none of which use anything besides plain data types. Given that, do you still think there might be a problem?

Reply:
That should be fine, then. My understanding is that 98 and 11 are ABI compatible for exactly the same code (same struct definitions), but there might no be a guarantee that you get the same STL (i.e. 98 and 11 might have different implementations of vector). So if you build with 98 and 11 on POD types, you should be fine.